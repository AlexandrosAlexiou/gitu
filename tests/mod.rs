use crate::helpers::{clone_and_commit, commit, key, key_code, run, TestContext};
use crossterm::event::KeyCode;
use std::fs;

mod helpers;

#[test]
fn no_repo() {
    let ctx = TestContext::setup_init(60, 20);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn help_menu() {
    let mut ctx = TestContext::setup_init(60, 20);
    ctx.update(&[key('h')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn fresh_init() {
    let mut ctx = TestContext::setup_init(60, 20);
    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn new_file() {
    let mut ctx = TestContext::setup_init(60, 20);
    run(ctx.dir.path(), &["touch", "new-file"]);
    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn unstaged_changes() {
    let mut ctx = TestContext::setup_init(60, 20);
    commit(ctx.dir.path(), "testfile", "testing\ntesttest");
    fs::write(ctx.dir.child("testfile"), "test\ntesttest").expect("error writing to file");

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn staged_file() {
    let mut ctx = TestContext::setup_init(60, 20);
    run(ctx.dir.path(), &["touch", "new-file"]);
    run(ctx.dir.path(), &["git", "add", "new-file"]);
    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn log() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "firstfile", "testing\ntesttest\n");
    commit(ctx.dir.path(), "secondfile", "testing\ntesttest\n");
    ctx.update(&[key('g'), key('l'), key('l')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn log_other() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "this-should-be-at-the-top", "");
    commit(ctx.dir.path(), "this-should-not-be-visible", "");

    ctx.update(&[key('g'), key('l'), key('l'), key('j'), key('l'), key('o')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn show() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "firstfile", "This should be visible\n");
    ctx.update(&[key('g'), key('l'), key('l'), key_code(KeyCode::Enter)]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn rebase_conflict() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "new-file", "hello");

    run(ctx.dir.path(), &["git", "checkout", "-b", "other-branch"]);
    commit(ctx.dir.path(), "new-file", "hey");

    run(ctx.dir.path(), &["git", "checkout", "main"]);
    commit(ctx.dir.path(), "new-file", "hi");

    run(ctx.dir.path(), &["git", "checkout", "other-branch"]);
    run(ctx.dir.path(), &["git", "rebase", "main"]);

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn merge_conflict() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "new-file", "hello");

    run(ctx.dir.path(), &["git", "checkout", "-b", "other-branch"]);
    commit(ctx.dir.path(), "new-file", "hey");

    run(ctx.dir.path(), &["git", "checkout", "main"]);
    commit(ctx.dir.path(), "new-file", "hi");

    run(ctx.dir.path(), &["git", "merge", "other-branch"]);

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn moved_file() {
    let mut ctx = TestContext::setup_clone(60, 20);
    commit(ctx.dir.path(), "new-file", "hello");
    run(ctx.dir.path(), &["git", "mv", "new-file", "moved-file"]);

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn hide_untracked() {
    let mut ctx = TestContext::setup_clone(60, 10);
    let mut config = ctx.state.repo.config().unwrap();
    config.set_str("status.showUntrackedFiles", "off").unwrap();
    run(ctx.dir.path(), &["touch", "i-am-untracked"]);

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn new_commit() {
    let mut ctx = TestContext::setup_clone(60, 10);
    commit(ctx.dir.path(), "new-file", "");

    ctx.update(&[key('g')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn push() {
    let mut ctx = TestContext::setup_clone(60, 10);
    commit(ctx.dir.path(), "new-file", "");

    ctx.update(&[key('P'), key('p')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn fetch_all() {
    let mut ctx = TestContext::setup_clone(60, 10);
    clone_and_commit(&ctx.remote_dir, "remote-file", "hello");

    ctx.update(&[key('f'), key('a')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}

#[test]
fn pull() {
    let mut ctx = TestContext::setup_clone(60, 10);
    clone_and_commit(&ctx.remote_dir, "remote-file", "hello");

    ctx.update(&[key('F'), key('p')]);
    insta::assert_snapshot!(ctx.redact_buffer());
}
