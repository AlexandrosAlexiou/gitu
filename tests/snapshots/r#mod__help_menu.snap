---
source: tests/mod.rs
expression: ctx.redact_buffer()
---
Buffer {
    area: Rect { x: 0, y: 0, width: 60, height: 20 },
    content: [
        "ðŸ¢’No branch                                                  ",
        "                                                            ",
        " Recent commits                                             ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€",
        "Help                Submenu                                 ",
        "g Refresh           h Help                                  ",
        "tab ToggleSection   b Branch                                ",
        "k p â†‘ SelectPreviou c Commit                                ",
        "j n â†“ SelectNext    f Fetch                                 ",
        "C-u HalfPageUp      l Log                                   ",
        "C-d HalfPageDown    F Pull                                  ",
        "y ShowRefs          P Push                                  ",
        "                    r Rebase                                ",
        "                    x Reset                                 ",
        "                    K Discard                               ",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 1, y: 0, fg: Rgb(216, 166, 87), bg: Rgb(80, 73, 69), underline: Reset, modifier: BOLD,
        x: 10, y: 0, fg: Reset, bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 0, y: 1, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 2, fg: Rgb(216, 166, 87), bg: Reset, underline: Reset, modifier: BOLD,
        x: 15, y: 2, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 8, fg: Rgb(80, 73, 69), bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 9, fg: Rgb(125, 174, 163), bg: Reset, underline: Reset, modifier: BOLD,
        x: 19, y: 9, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 9, fg: Rgb(125, 174, 163), bg: Reset, underline: Reset, modifier: BOLD,
        x: 40, y: 9, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 10, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 10, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 10, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 10, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 11, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 3, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 11, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 12, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 5, y: 12, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 12, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 12, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 13, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 5, y: 13, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 13, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 13, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 14, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 3, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 14, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 15, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 3, y: 15, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 15, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 15, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 16, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 16, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 17, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 17, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 18, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 19, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 19, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
